File: src/javascript/sorting/quickSort/quickSort.js
function swap(array, firstIndex, secondIndex) {
    let temp = array[firstIndex];
    array[firstIndex] = array[secondIndex];
    array[secondIndex] = temp;
}

function pivot(array, pivotIndex=0, endIndex=array.length-1) {
    let swapIndex = pivotIndex;
    for (let i = pivotIndex + 1; i <= endIndex; i++) {
        if (array[i] < array[pivotIndex]) {
        swapIndex++;
        swap(array, swapIndex, i);
        }  
    }
  swap(array, pivotIndex, swapIndex);
  
  return swapIndex;
}

function quickSort(array, left=0, right=array.length-1) {
    if(left < right) {
        let pivotIndex = pivot(array, left, right);
        quickSort(array, left, pivotIndex-1);
        quickSort(array, pivotIndex+1, right);
    }
}


let myUnsortedArray = [5, 1, 6, 2, 4, 3, 7];
quickSort(myUnsortedArray);
console.log(myUnsortedArray);

File: src/java1/sorting/quickSort/QuickSort.java
package java1.sorting.quickSort;

import java.util.Arrays;

public class QuickSort {

    private static void quickSort(int[] arr)  {
        quickSortHelper(arr, 0, arr.length-1);
    }

    private static void quickSortHelper(int[] arr, int left, int right) {
        if(left < right) {
            int pivot = pivot(arr, left, right);
            quickSortHelper(arr, left, pivot-1);
            quickSortHelper(arr, pivot+1, right);
        }
    }

    private static int pivot(int[] arr, int pivotIndex, int endIndex) {
        int swapIndex = pivotIndex;
        for (int i = pivotIndex+1; i <= endIndex; i++) {
            if(arr[i] < arr[pivotIndex]){
                swapIndex++;
                swap(arr, swapIndex, i);
            }
        }
        swap(arr, swapIndex, pivotIndex);

        return swapIndex;
    }

    private static void swap(int[] arr, int firstIndex, int secondIndex) {
        int temp = arr[firstIndex];
        arr[firstIndex] = arr[secondIndex];
        arr[secondIndex] = temp;
    }

    public static void main(String[] args) {
        int arr[] = {40, 50, 20, 0, -10, 30, 10};
        quickSort(arr);
        System.out.println(Arrays.toString(arr));
    }
}
